ctru-rs/src/services/cam.rs

```diff
@@ -898,36 +774,87 @@ pub trait Camera {
         &mut self,
         width: u16,
         height: u16,
-        transfer_unit: u32,
-        timeout: i64,
+        timeout: Duration,
     ) -> crate::Result<Vec<u8>> {
+        let transfer_unit = unsafe {
+            let mut buf_size = 0;
+            let r = ctru_sys::CAMU_GetMaxBytes(&mut buf_size, width as i16, height as i16);
+            if r < 0 {
+                Err(crate::Error::from(r))
+            } else {
+                Ok(buf_size)
+            }
+        }?;
+
         let screen_size = u32::from(width) * u32::from(width) * 2;
 
         let mut buf = vec![0u8; usize::try_from(screen_size).unwrap()];
 
-        self.set_transfer_bytes(
-            transfer_unit,
-            width.try_into().unwrap(),
-            height.try_into().unwrap(),
-        )?;
+        unsafe {
+            let r = ctru_sys::CAMU_SetTransferBytes(
+                self.port_as_raw(),
+                transfer_unit,
+                width as i16,
+                height as i16,
+            );
+            if r < 0 {
+                return Err(r.into());
+            }
+        };
 
-        self.activate()?;
```

Comment:
Maybe this can be useful for other use cases? For example, obtaining a continous view of the camera, but I haven’t looked at it directly. Also, it looks like the camera is never deactivated at the end of this function, did you forget?

Reply:
The libctru example of showing an active view of the camera uses other low level functions to do so including dealing with handles to detect interrupts. I think it may be better to have a separate function for previewing which is pretty much rendering an active video from the camera. [Here is the example.](https://github.com/devkitPro/3ds-examples/blob/master/camera/video/source/main.c)

Reply:
1. From the looks of it it seems they are actually just taking “multiple” pictures and flushing them to the framebuffer (with the only difference of checking the svchandle for a prolonged time rather than a single wait). Still, in our case, at least right now, the camera should be deactivated at the end of the `take_picture` function.